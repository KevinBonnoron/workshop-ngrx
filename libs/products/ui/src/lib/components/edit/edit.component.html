<h1>Edition d'un produit</h1>

<ng-container *ngIf="(product$ | async) as product">
  <products-ui-form [product]="product" (save)="onSave($event)"></products-ui-form>
</ng-container>

<button mat-raised-button color="warn" (click)="onDelete()">Supprimer</button>