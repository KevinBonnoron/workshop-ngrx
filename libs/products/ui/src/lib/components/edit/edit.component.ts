import { AsyncPipe, NgIf } from '@angular/common';
import { Component, inject } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { ActivatedRoute, Router } from '@angular/router';
import { ProductsService } from '@workshop/products-data-access';
import { Product } from '@workshop/shared';
import { switchMap } from 'rxjs';
import { FormComponent } from '../form/form.component';

@Component({
  standalone: true,
  imports: [MatButtonModule, FormComponent, NgIf, AsyncPipe],
  templateUrl: './edit.component.html',
})
export class EditComponent {
  private readonly router = inject(Router);
  
  private readonly activatedRoute = inject(ActivatedRoute);
  private readonly productsService = inject(ProductsService);

  readonly product$ = this.activatedRoute.params.pipe(switchMap(({ id }) => this.productsService.findOne(id)));

  onSave({ id, name }: Product) {
    this.productsService.update({ id, name }).subscribe(() => this.router.navigateByUrl('list'));
  }

  onDelete() {
    this.product$.pipe(
      switchMap((product) => this.productsService.remove(product.id))
    ).subscribe(() => this.router.navigateByUrl('list'));
  }
}
