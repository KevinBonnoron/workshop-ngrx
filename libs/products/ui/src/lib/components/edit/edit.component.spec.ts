import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ActivatedRoute, Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { ProductsService } from '@workshop/products-data-access';
import { Product } from '@workshop/shared';
import { of } from 'rxjs';
import { EditComponent } from './edit.component';

class BlankComponent {}

describe('EditComponent', () => {
  let fixture: ComponentFixture<EditComponent>;
  let component: EditComponent;
  let productsService: ProductsService;
  let router: Router;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule.withRoutes([{ path: 'list', component: BlankComponent }]),
        EditComponent
      ],
      providers: [
        { provide: ProductsService, useValue: {
          findOne: jest.fn(),
          update: jest.fn(),
          remove: jest.fn(),
        } },
        { provide: ActivatedRoute, useValue: { params: of({ id: 0 }) } },
      ]
    });

    fixture = TestBed.createComponent(EditComponent);
    component = fixture.componentInstance;
    productsService = TestBed.inject(ProductsService);
    router = TestBed.inject(Router);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should call endpoint when onSave is called', () => {
    const productServiceUpdateSpy = jest.spyOn(productsService, 'update').mockImplementation(() => of({ id: 0, name: 'UpdatedTest' }));
    const routerNavigateByUrlSpy = jest.spyOn(router, 'navigateByUrl').mockImplementation(() => Promise.resolve(true));

    const product: Product = { id: 0, name: 'UpdatedTest' };
    component.onSave(product);

    expect(productServiceUpdateSpy).toHaveBeenCalledWith(product);
    expect(routerNavigateByUrlSpy).toHaveBeenCalledWith('list');
  });

  it('should call endpoint when onDelete is called', () => {
    jest.spyOn(productsService, 'findOne').mockImplementation(() => of({ id: 0, name: 'Test' }));
    const productServiceRemoveSpy = jest.spyOn(productsService, 'remove').mockImplementation(() => of({ id: 0, name: 'Test' }));
    const routerNavigateByUrlSpy = jest.spyOn(router, 'navigateByUrl').mockImplementation(() => Promise.resolve(true));

    const product: Product = { id: 0, name: 'UpdatedTest' };
    component.onDelete();

    expect(productServiceRemoveSpy).toHaveBeenCalledWith(product.id);
    expect(routerNavigateByUrlSpy).toHaveBeenCalledWith('list');
  });

  it('should match snaptshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
