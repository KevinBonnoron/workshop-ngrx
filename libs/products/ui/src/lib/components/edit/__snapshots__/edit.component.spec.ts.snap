// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditComponent should match snaptshot 1`] = `
<ng-component
  activatedRoute={[Function Object]}
  product$={[Function Observable]}
  productsService={[Function Object]}
  router={[Function _Router]}
>
  <h1>
    Edition d'un produit
  </h1><button
    class="mdc-button mdc-button--raised mat-mdc-raised-button mat-warn"
    color="warn"
    mat-raised-button=""
    mat-ripple-loader-class-name="mat-mdc-button-ripple"
    mat-ripple-loader-uninitialized=""
  >
    <span
      class="mat-mdc-button-persistent-ripple"
    />
    <span
      class="mdc-button__label"
    >
      Supprimer
    </span>
    <span
      class="mat-mdc-focus-indicator"
    />
    <span
      class="mat-mdc-button-touch-target"
    />
  </button>
</ng-component>
`;
